src/util/xeval/evaluator.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use util::codec::{number, Datum};
+use super::{Result, Error};
+
+use std::collections::HashMap;
+use std::cmp::Ordering;
+use tipb::expression::{Expr, ExprType};
+
+/// `Evaluator` evaluates `tipb::Expr`.
+pub struct Evaluator {
+    // column_id -> column_value
+    row: HashMap<i64, Datum>,
+}
+
+impl Evaluator {
+    pub fn new(row: HashMap<i64, Datum>) -> Evaluator {
+        Evaluator { row: row }
+    }
+
+    /// Eval evaluates expr to a Datum.
+    pub fn eval(&self, expr: &Expr) -> Result<Datum> {
+        match expr.get_tp() {
+            ExprType::Int64 => self.eval_int(expr),
+            ExprType::Uint64 => self.eval_uint(expr),
+            // maybe we should use take here?
+            ExprType::String | ExprType::Bytes => Ok(Datum::Bytes(expr.get_val().to_vec())),
+            ExprType::ColumnRef => self.eval_column_ref(expr),
+            ExprType::LT => self.eval_lt(expr),
+            ExprType::LE => self.eval_le(expr),
+            ExprType::EQ => self.eval_eq(expr),
+            ExprType::NE => self.eval_ne(expr),
+            ExprType::GE => self.eval_ge(expr),
+            ExprType::GT => self.eval_gt(expr),
+            ExprType::NullEQ => self.eval_null_eq(expr),
+            ExprType::And => self.eval_and(expr),
+            ExprType::Or => self.eval_or(expr),
+            ExprType::Float32 | ExprType::Float64 => unimplemented!(),
+            _ => Ok(Datum::Null),
```

Comment:
@ngaut I don't think so. TiKV coprocessor can not scan table or index now. Maybe in the next pr.